src/tevutil.c

```diff
@@ -0,0 +1,690 @@
+    GXSetTevSwapModeTable_cached_init(GX_TEV_SWAP1, GX_CH_RED, GX_CH_GREEN, GX_CH_BLUE, GX_CH_RED);
+    GXSetTevSwapModeTable_cached_init(GX_TEV_SWAP2,GX_CH_RED, GX_CH_GREEN, GX_CH_BLUE, GX_CH_GREEN);
+    GXSetTevSwapModeTable_cached_init(GX_TEV_SWAP3, GX_CH_RED, GX_CH_GREEN, GX_CH_BLUE, GX_CH_BLUE);
+
+    GXSetTevKColor_cached_init(GX_KCOLOR0, color);
+    GXSetTevKColor_cached_init(GX_KCOLOR1, color);
+    GXSetTevKColor_cached_init(GX_KCOLOR2, color);
+    GXSetTevKColor_cached_init(GX_KCOLOR3, color);
+    GXSetFog_cached_init(GX_FOG_NONE, 0.0f, 100.0f, 0.1f, 20000.0f, color);
+
+    GXSetColorUpdate_cached_init(GX_TRUE);
+    GXSetAlphaUpdate_cached_init(GX_TRUE);
+    GXSetZCompLoc_cached_init(GX_TRUE);
+
+    for (stage = GX_TEVSTAGE0; stage < 0x10; stage++) {
+        GXSetTevSwapMode_cached_init(stage, GX_TEV_SWAP0, GX_TEV_SWAP0);
+        GXSetTevColorIn_cached_init(stage, GX_CC_ZERO, GX_CC_ZERO, GX_CC_ZERO, GX_CC_ZERO);
+        GXSetTevAlphaIn_cached_init(stage, GX_CA_ZERO, GX_CA_ZERO, GX_CA_ZERO, GX_CA_ZERO);
+        GXSetTevColorOp_cached_init(stage, GX_TEV_ADD, GX_TB_ZERO, GX_CS_SCALE_1, GX_TRUE, GX_TEVPREV);
+        GXSetTevAlphaOp_cached_init(stage, GX_TEV_ADD, GX_TB_ZERO, GX_CS_SCALE_1, GX_TRUE, GX_TEVPREV);
+        GXSetTevOrder_cached_init(stage, GX_TEXCOORD0, GX_TEXMAP0, GX_COLOR0A0);
+        GXSetTevKColorSel_cached_init(stage, GX_TEV_KCSEL_1);
+        GXSetTevKAlphaSel_cached_init(stage, GX_TEV_KASEL_1);
+    }
+    return 1;
+}
+
+void GXSetCullMode_cached(GXCullMode mode)
+{
+    if (mode != gxCache->cullMode) {
+        GXSetCullMode(mode);
+        gxCache->cullMode = mode;
+    }
+    return;
+}
+
+void GXSetCullMode_cached_init(GXCullMode mode, struct GXCache *cache)
+{   
+    GXSetCullMode(mode);
+    gxCache->cullMode = mode;
+    return;
+}
+
+void GXSetBlendMode_cached(GXBlendMode type, GXBlendFactor src_factor, GXBlendFactor dst_factor, GXLogicOp op)
+{
+    if (type == GX_BM_LOGIC) {
+        if (
+            ( (gxCache->blendMode).type != GX_BM_LOGIC) || ((gxCache->blendMode).op != op )
+        ) {
+            GXSetBlendMode(GX_BM_LOGIC, (gxCache->blendMode).src_factor, (gxCache->blendMode).dst_factor, op);
+            (gxCache->blendMode).type = GX_BM_LOGIC;
+            (gxCache->blendMode).op = op;
+        }
+    } else if (
+        ( (gxCache->blendMode).type != type || (gxCache->blendMode).src_factor != src_factor ) ||
+        (gxCache->blendMode).dst_factor != dst_factor
+    )
+    {
+        GXSetBlendMode(type, src_factor, dst_factor, (gxCache->blendMode).op);
+        (gxCache->blendMode).type = type;
+        (gxCache->blendMode).src_factor = src_factor;
+        (gxCache->blendMode).dst_factor = dst_factor;
+    }
+    return;
+}
+
+void GXSetBlendMode_cached_init(GXBlendMode type, GXBlendFactor src_factor, GXBlendFactor dst_factor, GXLogicOp op)
+{
+    GXSetBlendMode(type, src_factor, dst_factor, op);
+    (gxCache->blendMode).type = type;
```

Comment:
These parentheses here are probably not necessary. If it still matches, you can get rid of them.
`gxCache->blendMode.type = type;`

Reply:
I confirmed It's still matching.
Fixed pick up codes.
And fix similar codes.